doc/design/zarr-support.md

```diff
@@ -0,0 +1,100 @@
+   a zarr-compliant software can navigate the zarr metadata/structure using relative path rules.
+3. Blob url should continue to point to a single location (i.e. the prefix ending with `/` common to all keys for zarr files) in an s3 bucket instead of a collection of files.
+4. The `etag` connected with the zarr file should be a tree-hash of dandi-etags of the folder.
+
+Storing each sub-file of a zarr "dataset" as an asset seems like an overkill without any specific gain at this point.
+
+Given these considerations here are questions for implementation
+1. Is there a way to upload a folder to a given prefix using an API key without having to create 100k signed URLs?
+1. Should the tree structure be stored somewhere so that diffs can be ascertained?
+   * Assuming 64 bytes of checksum, 64 bytes for path (optimistic), and 1million files (pessimistic), that's 128MB of checksum info in the DB. Plausible for upload, but probably not long term.
+1. Given that each zarr file may contain 100k+ files, how will dandi-cli handle alterations?
+
+# Requirements
+
+1. zarr files are stored in a "directory" in S3.
+1. Each zarr file corresponds to a single Asset.
+1. The CLI uses some kind of tree hashing scheme to compute a checksum for the entire zarr file. The API verifies this checksum _immediately_ after upload; it's not good enough to download the entire zarr file to calculate it after upload.
+1. The system can theoretically handle zarr files with ~1 million subfiles, each of size 64 * 64 * 64 bytes ~= 262 kilobytes.
+
+# Implementation
+
+## Hash function for checksums
+The hash function used to generate the checksum for the zarr file must have these characteristics:
+
+1. It must be based on the dandi-etag function of the subfile contents, and the path of the subfile in the zarr file.
+1. A well-defined ordering of the subfiles in the zarr file must exist.
+The checksum must be computed on the subfiles in this order.
+1. It must be applicable on one subfile at a time.
+1. It must be able save its state between subfiles.
+
+**TODO** what is this function?
+
+A simple scheme that I think would work:
+
+* initial value is `sha256("{etag}:{path}")` for the first subfile
+* the next value is `sha256("{prev_sha256}:{etag}:{path}")`, ad infinitum
+
+We could also do something more complicated with merkle trees or something, but it would be much more obnoxious to compute.
+
+## Upload flow
+
+### Before upload (technically optional)
+1. CLI calculates the checksum of the zarr file.
+1. CLI queries the API (**TODO** URL?) if the checksum has already been uploaded.
+1. If so, it proceeds with the already uploaded zarr file and skips upload entirely.
+
+### Upload
+1. CLI queries the API (**TODO** URL?) with the checksum of the zarr file to initiate an upload. The API creates an upload UUID, records the checksum, and initializes a "running checksum" to `null`.
```

Comment:
the same `/uploads/initiate`, `contentSize` - sum of all files sizes (although not used for checksum compute), but either we have a different name for the checksum or provide an explicit extra option to specify the `format='zarr'` (defaulting to `'file'` as that is what we support now).

Reply:
I left the actual API definition for later so we can discuss it after we agree on the requirements and that the overall implementation will meet those requirements.